File: server/provenance/canonical.ts
/**
 * RFC 8785 JSON Canonicalization Scheme -- the TypeScript half.
 *
 * This must produce byte-identical output to `ml-engine/provenance/canonical.py` for
 * every input. If the two drift, a record sealed by one side fails verification on the
 * other and the console reports TAMPERED for a record nobody touched -- the fastest way
 * to make operators stop trusting an integrity tool. `server/provenance/canonical.test.ts`
 * pins the shared vectors.
 *
 * JavaScript happens to fit JCS almost exactly, which is not a coincidence: the spec is
 * built on ECMAScript's own serialisation rules.
 *
 *  - `JSON.stringify` on a number *is* `Number::toString`, which is what the spec cites.
 *  - `JSON.stringify` on a string produces exactly the spec's minimal escaping: short
 *    forms for the five named controls, `\u00xx` for the rest, non-ASCII left as UTF-8.
 *  - `Array.prototype.sort()` with no comparator orders by UTF-16 code unit, which is
 *    the spec's key ordering.
 *
 * So the work here is structural: sort keys, emit no whitespace, and refuse the values
 * JSON cannot represent rather than silently coercing them.
 */

export type JsonValue = null | boolean | number | string | JsonValue[] | { [key: string]: JsonValue };

export function canonicalize(value: unknown): string {
  if (value === null) return 'null';

  const type = typeof value;

  if (type === 'boolean') return value ? 'true' : 'false';

  if (type === 'number') {
    const numeric = value as number;
    if (!Number.isFinite(numeric)) {
      // RFC 8785 section 3.2.2.3: NaN and Infinity have no JSON representation. Coercing
      // them to null would let two different records share a digest.
      throw new TypeError('NaN and Infinity cannot be canonicalised');
    }
    return JSON.stringify(numeric);
  }

  if (type === 'string') return JSON.stringify(value);

  if (type === 'bigint') {
    throw new TypeError('BigInt cannot be canonicalised: it has no JSON number representation');
  }

  if (Array.isArray(value)) {
    return `[${value.map((item) => canonicalize(item)).join(',')}]`;
  }

  if (type === 'object') {
    const record = value as Record<string, unknown>;
    // Default sort is UTF-16 code unit order, which is what the spec requires.
    const keys = Object.keys(record).sort();
    const parts: string[] = [];
    for (const key of keys) {
      const inner = record[key];
      // Match JSON.stringify: an undefined member is omitted, not encoded as null.
      if (inner === undefined) continue;
      parts.push(`${JSON.stringify(key)}:${canonicalize(inner)}`);
    }
    return `{${parts.join(',')}}`;
  }

  throw new TypeError(`${type} is not JSON-serialisable and cannot be canonicalised`);
}

export function canonicalBytes(value: unknown): Buffer {
  return Buffer.from(canonicalize(value), 'utf8');
}

File: ml-engine/provenance/canonical.py
"""RFC 8785 JSON Canonicalization Scheme.

Tamper detection is only as good as the serialisation it hashes. An ad-hoc scheme --
joining fields with a separator, or `json.dumps` with default settings -- breaks in ways
that are easy to overlook and fatal in practice:

* **Separator injection.** Joining fields with ``||`` means a prediction of
  ``"STOP||0.99"`` collides with a different record. The digest stops being a function of
  the *structure*.
* **Cross-implementation drift.** Python renders ``0.1 + 0.2`` and large integers
  differently from JavaScript. A record sealed by the Python engine would then fail
  verification in the Node gateway, producing false TAMPERED alerts -- which is how an
  integrity system gets switched off.
* **Key ordering and escaping.** Two semantically identical objects must produce
  identical bytes.

RFC 8785 (JCS) solves all three: keys sorted by UTF-16 code unit, no insignificant
whitespace, minimal string escaping, and numbers rendered by the ECMAScript
``Number::toString`` algorithm. Implementing the same scheme in Python and TypeScript is
what lets either side seal a record and the other verify it.
"""

from __future__ import annotations

import math
import re
from typing import Any

_ESCAPES = {
    '"': '\\"',
    "\\": "\\\\",
    "\b": "\\b",
    "\f": "\\f",
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
}

_EXPONENT = re.compile(r"^(-?)(\d)(?:\.(\d+))?e([+-])(\d+)$")


def serialize_number(value: float | int) -> str:
    """Render a number per the ECMAScript ``Number::toString`` algorithm.

    This is the fiddly part of JCS. Python's ``repr`` agrees with JavaScript for most
    doubles because both emit the shortest round-tripping representation, but the
    exponent formatting and the integer/float boundary differ, so those are normalised
    explicitly.
    """
    if isinstance(value, bool):  # bool is a subclass of int; never let it through here
        raise TypeError("bool is not a JSON number")

    if isinstance(value, int):
        return str(value)

    if math.isnan(value) or math.isinf(value):
        raise ValueError("NaN and Infinity cannot be canonicalised (RFC 8785 section 3.2.2.3)")

    if value == 0:
        # JavaScript renders -0 as "0".
        return "0"

    # An integral double renders without a fractional part in JavaScript.
    if value.is_integer() and abs(value) < 1e21:
        return str(int(value))

    text = repr(float(value))

    if "e" in text or "E" in text:
        text = text.replace("E", "e")
        match = _EXPONENT.match(text)
        if match:
            sign, lead, fraction, exp_sign, exp_digits = match.groups()
            mantissa = f"{lead}.{fraction}" if fraction else lead
            exponent = int(exp_digits)
            # ECMAScript uses decimal notation for exponents in [-7, 21).
            if exp_sign == "+" and exponent < 21:
                return f"{sign}{float(text):.0f}" if float(text).is_integer() else f"{sign}{float(text)!r}"
            text = f"{sign}{mantissa}e{'+' if exp_sign == '+' else '-'}{exponent}"

    return text


def serialize_string(value: str) -> str:
    """Minimal JSON string escaping per RFC 8785 section 3.2.2.2."""
    out = ['"']
    for char in value:
        escape = _ESCAPES.get(char)
        if escape is not None:
            out.append(escape)
        elif ord(char) < 0x20:
            out.append(f"\\u{ord(char):04x}")
        else:
            out.append(char)
    out.append('"')
    return "".join(out)


def canonicalize(value: Any) -> str:
    """Produce the canonical JSON text for `value`."""
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, str):
        return serialize_string(value)
    if isinstance(value, (int, float)):
        return serialize_number(value)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(canonicalize(item) for item in value) + "]"
    if isinstance(value, dict):
        # RFC 8785 sorts by UTF-16 code unit. Python sorts str by code point, which
        # differs only for characters above the BMP; encoding to UTF-16 makes it exact.
        items = sorted(value.items(), key=lambda kv: str(kv[0]).encode("utf-16-be"))
        return "{" + ",".join(f"{serialize_string(str(k))}:{canonicalize(v)}" for k, v in items) + "}"

    raise TypeError(f"{type(value).__name__} is not JSON-serialisable and cannot be canonicalised")


def canonical_bytes(value: Any) -> bytes:
    return canonicalize(value).encode("utf-8")


__all__ = ["canonical_bytes", "canonicalize", "serialize_number", "serialize_string"]
